### PR TITLE
Enable local relative path copy of images.

### PR DESCRIPTION
**Pull request by didrocks**
As all images were imported from gdoc (and so remote), the case of local images
from markdown files weren't handled.
Ensure we can now fetch images remotely (from gdoc, drive or remote reference
from a markdown file) or locally. We compute relative path from the source
file itself for local images and reuse it when running "update".

### CODE REVIEW

**Comment by x1ddos** [comment]
@cassierecher what do you think about this one? not sure if it makes any difference to you, but maybe it does.

my personal concern is security. what if I create a markdown with an image referencing something like `/etc/passwd`.

**Comment by didrocks** [comment]
`/etc/passwd` isn't an issue per itself, but I guess you meant something like `/etc/shadow` which has the hash passwords values. This one isn't readable to the world anyway. The `claat` tool will only have the permissions that the user allowed it to have. I understand your concern about exposing by error things from your system you don't want. (but for sensitive items, it means you gave the claat tool those priviledges).

If we want to prevent people from doing this, we can maybe we can filter by extension (or mimetype) to allow a certain list of allowed items? What do you think about this?

**Comment by cassierecher** [comment]
@x1ddos The overall strategy and implementation LGTM, but I agree that there's potential security implications. I defer to your expertise on that; I haven't done much in-depth security work.

File: claat/fetch.go [hunk 1]
```diff
@@ -227,22 +228,42 @@ func fetchDriveFile(id string, nometa bool) (*resource, error) {
...
+	var b []byte
+	var ext string
+	u, err := url.Parse(imgURL)
 	if err != nil {
 		return "", err
 	}
-	defer res.Body.Close()
-	b, err := ioutil.ReadAll(res.Body)
+	if u.Host == "" {
+		imgURL = path.Join(path.Dir(codelabSrc), imgURL)
```
Comment (x1ddos) [inline claat/fetch.go]: how about prepending "/" to imgURL here, to make sure it doesn't go like "../../../etc/passwd" or any other sensitive content.

then, path.Join should take care of cleaning up the final result anyway.

Comment (didrocks) [inline claat/fetch.go]: I think that restricting elements to a subpath from the codelab directory is a nice way to mitigate those issues.

However, prepending a / to imgURL will not achieve this (I just tried and checked golang stdlib): `path.Join` is cleaning each elements of the `Join` (apart from the first one), and thus, strips any prepended or appended /. Thus, `path.Join("bar", "/../foo")` is equivalent to `path.Join("bar", "../foo")`.
I'm going to use some absolute path method to ensure we capture the asset as being in the tree of the codelab directory.

**Comment by didrocks** [comment]
In light of the previous element, I think that ensuring the asset is a children of the codelab directory should be enough, what do you think?
The issue with checking mimetype is that it's ok for most of the types, but svgs are thus "text/plain; charset=utf-8", which will require us either to not allow them (as all sensitive data will match the same mimetype) or make that check useless, thoughts?

**Comment by x1ddos** [comment]
Sure, let's do just the path cleanup.
Also, note that you should probably be using `filepath`, not the `path` package, when working with file system.

**Comment by didrocks** [comment]
Sounds ok! I did thus push the asset being constrained in the corresponding codelab {;sub-}directory.
Also, changed the path -> filepath (never noticed that filepath had a Join() method)

Tell me if you prefer that I squash that in a single commit.

File: claat/fetch.go [hunk 2]
```diff
@@ -227,22 +228,49 @@ func fetchDriveFile(id string, nometa bool) (*resource, error) {
...
+	var b []byte
+	var ext string
+	u, err := url.Parse(imgURL)
 	if err != nil {
 		return "", err
 	}
-	defer res.Body.Close()
-	b, err := ioutil.ReadAll(res.Body)
+	if u.Host == "" {
+		cdir, err := filepath.Abs(path.Dir(codelabSrc))
```
Comment (x1ddos) [inline claat/fetch.go]: I may sound a bit paranoid but it would be nice if you extracted this part, sanitizing local file path, into a separate function and wrote a test with all possible edge cases. I would also use `testing/quick` in addition to manual test cases.

File: claat/fetch.go [hunk 3]
```diff
@@ -227,22 +228,49 @@ func fetchDriveFile(id string, nometa bool) (*resource, error) {
...
+		if err != nil {
+			return "", err
+		}
+		imgURL = filepath.Join(path.Dir(codelabSrc), imgURL)
```
Comment (x1ddos) [inline claat/fetch.go]: This should be `filepath.Dir`, not `path.Dir`, right.
Same above.

**Comment by didrocks** [comment]
The latest commits should address and include your suggestions of factoring this out.
It does add tests for the functions (both manual ones with edge cases and fuzzy testing).

Let me know if this is what you had in mind. I'm happy to squash the commits if desired.

File: claat/fetch_test.go [hunk 4]
```diff
@@ -16,12 +16,17 @@ package main
...
+	"fmt"
 	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
 	"strings"
 	"testing"
 
+	"path/filepath"
+
+	"testing/quick"
```
Comment (x1ddos) [inline claat/fetch_test.go]: Could you remove empty lines from the standard pkg imports? That way they will all line up nicely.

**Comment by x1ddos** [comment]
@didrocks sorry to keep this PR waiting. Just one nit in the package imports in fetch_test.go and I think we can merge. 

Please, do squash commits if possible.
Thanks!

File: claat/fetch_test.go (hunk 4 above)
Comment (didrocks) [inline claat/fetch_test.go]: You're right, seems like my go-import doesn't always act as nicely as desired. Fixing it!

**Comment by didrocks** [comment]
No worry at all! Maintaining myself a lot of opensource repo, I'm know what this means on your side :)

Made the latest fix, resquashed in one commit (updating its description) and rerun the tests. All should be good now! Thanks for the review :)

**Comment by x1ddos** [comment]
Thanks!
